plugins/mongodb/mongodbembed.cpp

```diff
@@ -0,0 +1,2302 @@
+                else if (stricmp(optName, "user")==0)
+                    user = val;
+                else if (stricmp(optName, "password")==0)
+                    password = val;
+                else if (stricmp(optName, "database")==0)
+                    databaseName = val;
+                else if (stricmp(optName, "collection")==0)
+                    collectionName = val;
+                else if (stricmp(optName, "batchSize")==0)
+                    batchSize = atoi(val);
+                else
+                    failx("Unknown option %s", optName.str());
+            }
+        }
+
+        if (port != 0)
+            connectionString.appendf("mongodb://localhost:%i", port); // Connection to a local MongoDB instance
+        else
+            connectionString.appendf("mongodb+srv://%s:%s@%s/?retryWrites=true&w=majority", user, password, server); // Connection to MongoDB Atlas server
+        std::shared_ptr<MongoDBQuery> ptr(new MongoDBQuery(databaseName, collectionName, batchSize));
+        query = ptr;
+
+        std::call_once(CONNECTION_CACHE_INIT_FLAG, configure, mongocxx::uri{connectionString.str()});
+    }
+
+    /**
+     * @brief Destroy the MongoDBEmbedFunctionContext object.
+     */
+    MongoDBEmbedFunctionContext::~MongoDBEmbedFunctionContext()
+    {
+    }
+
+    /**
+     * @brief Builds an ECL dataset from the result documents of MongoDB query
+     * 
+     * @param _resultAllocator Used for building the ECL dataset by the engine.
+     * @return IRowStream* Stream to ECL dataset handed back to the engine.
+     */
+    IRowStream * MongoDBEmbedFunctionContext::getDatasetResult(IEngineRowAllocator * _resultAllocator)
+    {
+        Owned<MongoDBRowStream> mdbRowStream;
+        mdbRowStream.setown(new MongoDBRowStream(_resultAllocator, query));
+        return mdbRowStream.getLink();
+    }
+
+    /**
+     * @brief Builds an ECL row from the result documents of MongoDB query
+     * 
+     * @param _resultAllocator Used for building the ECL row by the engine.
+     * @return byte* Pointer to ECL row handed back to the engine.
+     */
+    byte * MongoDBEmbedFunctionContext::getRowResult(IEngineRowAllocator * _resultAllocator)
+    {
+        Owned<MongoDBRowStream> mdbRowStream;
+        mdbRowStream.setown(new MongoDBRowStream(_resultAllocator, query));
+        return (byte *)mdbRowStream.getLink();
+    }
+
+    /**
+     * @brief Currently Unimplemented.
+     * 
+     * @param rowBuilder 
+     * @return size32_t 
+     */
+    size32_t MongoDBEmbedFunctionContext::getTransformResult(ARowBuilder & rowBuilder)
+    {
+        // To Do
+        size32_t num = 0;
+        return num;
+    }
+
+    /**
+     * @brief Returns the success of the operation.
+     * 
+     * @return true Result Rows has a length greater than 0.
+     * @return false If no result was returned and the result rows are empty.
+     */
+    bool MongoDBEmbedFunctionContext::getBooleanResult() 
```

Comment:
MongoDB does not return scalar values it only returns JSON objects so there isn't necessarily a next scalar. The order of these documents and their fields is not guaranteed, so an ECL user trying to get a single scalar field from the document will not always get the same value. For these scalar return types I was able to find a few datatypes that could represent something useful about the MongoDB query rather than returning a random scalar value from the MongoDB document/JSON object.

Reply:
Fair enough.
So it sounds but it sounds like this operation should be UNSUPPORTED

Reply:
Yeah that makes a lot of sense. The query throws an exception if it fails, so returning the boolean is redundant. I changed the other scalar types to unimplemented as well. For the insertMany function it returns a count of the inserted documents so I return that as a result row rather than a scalar now. I added two more standard record layouts to the ecllib file to avoid someone messing up on the name of the field that is returned.